pkg/envfuncs/provider_funcs.go

```diff
@@ -0,0 +1,242 @@
+)
+
+type clusterNameContextKey string
+
+// Following placeholder interface types are defined purely to make the pointer receiver work seemlessly
+// for the generic handlers. Without this, making the pointer receiver is really painful or we have to
+// turn providers into value receivers and that doesn't really work well for the current implementation
+// we have for kind provider. xref: https://t.ly/GDxOB
+
+// Every time you add a new provider that implements the support.E2EClusterProvider interface, that
+// needs to be listed here in this interface type with properly or'ing that with the existing
+// values.
+//
+//	type provider interfae {
+//			kind.Cluster | k3d.Cluster | kwok.Cluster
+//	}
+type provider interface {
+	kind.Cluster
+}
+
+type providerRef[T any] interface {
+	*T
+	support.E2EClusterProvider
+}
+
+// GetClusterFromContext helps extract the E2EClusterProvider object from the context.
+// This can be used to setup and run tests of multi cluster e2e Prioviders.
+func GetClusterFromContext(ctx context.Context, clusterName string) (support.E2EClusterProvider, bool) {
+	c := ctx.Value(clusterNameContextKey(clusterName))
+	if c == nil {
+		return nil, false
+	}
+	cluster, ok := c.(support.E2EClusterProvider)
+	return cluster, ok
+}
+
+// Deprecated: This handler has been deprecated in favor of GetClusterFromContext
+func GetKindClusterFromContext(ctx context.Context, clusterName string) (*kind.Cluster, bool) {
+	provider, ok := GetClusterFromContext(ctx, clusterName)
+	if ok {
+		return provider.(*kind.Cluster), ok
+	}
+	return nil, ok
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateCluster which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindCluster(clusterName string) env.Func {
+	return CreateCluster[kind.Cluster](clusterName)
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateClusterWithConfig which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindClusterWithConfig(clusterName, image, configFilePath string) env.Func {
+	return CreateClusterWithConfig[kind.Cluster](clusterName, configFilePath, kind.WithImage(image))
+}
+
+// Deprecated: This handler has been deprecated in favor of DestroyCluster
+func DestroyKindCluster(name string) env.Func {
+	return DestroyCluster(name)
+}
+
+// Deprecated: This handler has been deprecated in favor of ExportClusterLogs
+func ExportKindClusterLogs(name, dest string) env.Func {
+	return ExportClusterLogs(name, dest)
+}
+
+// CreateCluster returns an env.Func that is used to
+// create an E2E provider cluster that is then injected in the context
+// using the name as a key.
+//
+// NOTE: the returned function will update its env config with the
+// kubeconfig file for the config client.
+func CreateCluster[T provider, P providerRef[T]](clusterName string) env.Func {
```

Comment:
What if `CreateCluster` takes an E2EClusterProvider as a type constraint  as follows?

```go
func CreateCluster[P E2EClusterProvider](provider P, clusterName string) env.Func { 
...
}
```

I think it will work simpler, right ? Since your type constraint will be an interface you should be able to use type parameter that only needs to implement the type constraint as follows:

```go
func CreateKindCluster(clusterName string) env.Func {
	return CreateCluster[kind.Cluster](p kind.Cluster, clusterName)
}
```

I could be wrong. But it's worth looking into if it helps simplifying the implementation.

Reply:
@vladimirvivien  this is what me and @jbpratt  were discussing earlier 

xref: https://github.com/kubernetes-sigs/e2e-framework/pull/246#issuecomment-1586079640
xref: https://github.com/kubernetes-sigs/e2e-framework/pull/246#issuecomment-1585486410

If we are to use the interface directly, like this, it will work as value receiver, which will break our workflow where we set the kubeconfig files to the context of the cluster provider as part of the cluster create. We have to make that as pointer receiver and the hack I have put up right now is the only way to do that as far as I know.

Reply:
@vladimirvivien  I was speaking to @merovius on a Gopher channel and Axel suggested something that is much simper that what I have been doing. https://go.dev/play/p/C7RbbQYQzyN

This seem to do the job. May be this will simplify things for us.

Reply:
And the other suggestion that made a lot of sense was to just avoid generics entirely and keep it simple. Which I think would be the best thing to do here since generics is not really making this any better. cc @jbpratt

Reply:
All right.. Let me refactor and cleanup the generics and use a simple approach like we had before